src/msw/desktopenv.cpp

```diff
@@ -0,0 +1,56 @@
+/////////////////////////////////////////////////////////////////////////////
+// Name:        src/msw/desktopenv.cpp
+// Purpose:     wxDesktopEnv
+// Author:      Igor Korot
+// Modified by:
+// Created:     17/11/15
+// Copyright:   (c) wxWidgets team
+// Licence:     wxWindows licence
+/////////////////////////////////////////////////////////////////////////////
+
+// ============================================================================
+// declarations
+// ============================================================================
+
+// ----------------------------------------------------------------------------
+// headers
+// ----------------------------------------------------------------------------
+
+// For compilers that support precompilation, includes "wx.h".
+#include "wx/wxprec.h"
+
+#ifdef __BORLANDC__
+    #pragma hdrstop
+#endif
+
+#include "wx/desktopenv.h"
+
+// ============================================================================
+// implementation
+// ============================================================================
+
+wxDesktopEnv::wxDesktopEnv() : wxDesktopEnvBase()
+{
+}
+
+bool wxDesktopEnv::MoveFileToRecycleBin(const wxString &fileName)
+{
+    bool result = false;
+    if( wxDesktopEnvBase::MoveFileToRecycleBin( fileName ) )
+    {
+        SHFILEOPSTRUCT fileOp;
+        wxZeroMemory( fileOp );
+        fileOp.hwnd = NULL;
+        fileOp.wFunc = FO_DELETE;
+        wxString temp = fileName + '\0';
+        fileOp.pFrom = temp.c_str();
+        fileOp.pTo = NULL;
+        fileOp.fFlags = FOF_ALLOWUNDO | FOF_NOERRORUI | FOF_NOCONFIRMATION | FOF_SILENT;
+        int res = SHFileOperation( &fileOp );
+        if( res != 0 )
+           wxLogSysError( _ ( "Failed to move file '%s' to Recycle Bin" ), fileName.c_str() );
+        else
+            result = true;
+    }
+    return result;
+}
```

Comment:
My suggestions:
1. Do not use `wxString::c_str()`, use `t_str()` instead when setting the structure field and use just `fileName` when reporting an error. 
2. Perform a check assuring `fileName` is an absolute path (otherwise `FOF_ALLOWUNDO` is ignored).
3. Check whether using `wxLogSysError()` is correct here: The docs says do not use GetLastError() for `wxLogSysError()` with the returned value but perhaps it is incorrect even when checking the thread's last error code.
4. Fix code style issues (whitespace, ampersand placement, using const for `res`...)

BTW, I am not sure if that extra ending `0` is handled properly here...

Reply:
@PBfordev ,
1. I don't know what t_str() does, but it looks like it works, so changed it.
2. Added the check for IsAbsolute() to the parent class - it doesn't make sense to check it in MSW only.
3. I'm not sure where did you find me using GetLastError()?
4. const added.
5. Can you elaborate the comment about '\0'?

Reply:
**Ad 1.** Sorry, seems I was wrong about `c_str()`. For some reason I always thought it returns a narrow char pointer even in Unicode builds, which is not true. Additionally, `fn_str()` is probably the `wxString` method intended for using with file paths but I guess it is all the same now and `c_str()` is fine.

**Ad 2.** MSW requires an absolute path for undo to work but perhaps other OSes work differently (I am MSW only)?

**Ad 3.** I thought `wxLogSysError()` uses` ::GetLastError()` (via `wxSysCode()`) and appends the system message for the last error to the one passed to `wxLogSysError()`?

**Ad 5.** I meant whether `fileOp.pFrom` really contains two trailing `0`s, after being assigned its value from a `wxString`.

Reply:
FWIW we must indeed use `t_str()` here to make this code compile in `wxUSE_UNICODE_UTF8` build.

I also agree with the other PBfordev's comments.